Client/Application/AppDelegate.swift

```diff
@@ -210,6 +217,16 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
         return true
     }
 
+    fileprivate func updateSyncScratchpad() {
+        // We need to associate the fxaDeviceId with sync;
+        // We can do this anything after the first time we account.advance()
+        // but before the first time we sync.
+        if let deviceRegistration = profile?.getAccount()?.deviceRegistration,
+            let scratchpadPrefs = profile?.prefs.branch("sync.scratchpad") {
+            scratchpadPrefs.setString(deviceRegistration.toJSON().stringValue()!, forKey: PrefDeviceRegistration)
```

Comment:
This code has gone through several revisions (e.g., #2893), but I think it's fundamentally wrong. FxALoginHelper should not be reaching into `sync.scratchpad`.

If a sync is in progress, this change will get wiped out. If a sync isn't in progress, then this work shouldn't be needed — the right channel for communication between the FxA state machine and the Sync state machine is the result of `advance`.

I suggest:

- Making sure the device registration is part of the account objects. It probably already is.
- Having the Sync code detect, based on that, whether something material has changed about the account that would warrant a re-sync of e.g., clients.
- Having the clients synchronizer decide if it needs to reupload the client record for that reason.
- In the notification handler above, go through `profile.syncManager` to trigger a sync of clients.

Reply:
@rnewman 

So understand poking things into the scratchpad isn't the right approach. That hack was trying to pass the `deviceRegistration.id` to `ClientsSynchronizer.getOurClientRecord()`, so how do we achieve that now?

I note that the [`authState`][authState] passes to `toReady` is always a `FirefoxAccountSyncAuthState`. Wondering about allowing access to the account or deviceRegistration, so I can put the `fxaDeviceId` into the scratch pad in [`toReady`][toReady].

Unfortunately, [more code][hashedFxADeviceId] has been written with the assumption that `fxaDeviceId` is in the scratchpad.

As to the rest of your list: 

- [x]: Making sure the device registration is part of the account objects.
- [ ]: Having the Sync code detect, based on that, whether something material has changed about the account that would warrant a re-sync of e.g., clients. — I think this is [`NotificationFirefoxAccountDeviceRegistrationUpdated`][NotificationFirefoxAccountDeviceRegistrationUpdated] —  or are you suggesting something in [`toReady`][toReady] as above?
- [ ]: Having the clients synchronizer decide if it needs to reupload the client record for that reason. — Is there any harm in uploading the client record everytime?
- [ ]: In the notification handler above, go through profile.syncManager to trigger a sync of clients. — ok.



[authState]: https://github.com/mozilla-mobile/firefox-ios/blob/master/Providers/Profile.swift#L1103
[toReady]: https://github.com/mozilla-mobile/firefox-ios/blob/master/Sync/SyncStateMachine.swift#L139
[hashedFxADeviceId]: https://github.com/mozilla-mobile/firefox-ios/blob/master/Sync/SyncStateMachine.swift#L900
[NotificationFirefoxAccountDeviceRegistrationUpdated]: https://github.com/mozilla-mobile/firefox-ios/blob/0120d8a69466b5c097a68cef54f092b25f8bbee9/Account/FirefoxAccount.swift#L345

Reply:
It's OK for the device ID to be in the scratchpad — it's just `SyncStateMachine` that should put it there, providing a single place to write it, persist it, and then recognize that it's changed (either from not existing to existing, or if it changes for some other reason).

Re FxA notifications: don't drive parts of the Sync code itself with notifications. It's 'incorrect' in that it's coupling that doesn't make sense, but it also would introduce a second parallel control flow that could occur at any point during a sync, which is hard to get right. These two state machines exist to manage the very significant complexity that occurs during sign-in and the start of a sync; we should not try to circumvent that.

Instead, in the `SSM` we can look at the account state we're given and decide — by looking at our last scratchpad — if something has changed. For example, if the device ID has changed from nothing to a valid device ID, then we can [reset the clients synchronizer](https://github.com/mozilla-mobile/firefox-ios/blob/master/Sync/SyncStateMachine.swift#L906) as we advance, ensuring that we'll upload our own record again.

This will behave correctly, too: e.g., if an error occurs that causes us not to sync clients, we'll remember to do it next time we sync.

Re uploading the client record _every time_ we sync: it's needless load on the server, and it's needless load on other clients, too. Every other client will have to download it each time. Not to mention that it causes risk of conflict for sending tabs and other commands, and a year from now some engineer is going to say "oh, why are we doing this? that's silly, I'll fix it", and will break whatever fix you introduce here.